### ISSUE TITLE
gh-ost migration: [error] close sync with err: sync is been closing...

### ISSUE THREAD

**Issue by ChinaVogel** [github/gh-ost#388]
Hello gh-ost team,
I get this error when running a migration with gh-ost:
binlogstreamer.go:47: [error] close sync with err: sync is been closing...

the migration seems to run fine (at least the table got changed ;) and all the content is there)
But this such an error make me feel uncomfortable using it on production.

System: macOS Sierra 10.12.3
Installed: just running the downloaded executable: 
```
> /gh-ost --version
1.0.35
```

DB System:
```
> mysql --version
mysql Ver 15.1 Distrib 10.2.4-MariaDB, for osx10.12 (x86_64) using readline 5.1
```

Is this more a warning? I did not look into source code of gh-ost so far. I got this message also for other fields and other migrations.

regards

Here the whole verbose output (it is a test run with root on a dev machine, that is why I use high secure no pwd root... )
```
./gh-ost --database="excelsior_development" --table="attachments" --user="root" --alter="MODIFY filesize BIGINT DEFAULT 0" --switch-to-rbr --allow-on-master --initially-drop-ghost-table --verbose --execute
2017-03-27 11:15:24 INFO starting gh-ost 1.0.35
2017-03-27 11:15:24 INFO Migrating `excelsior_development`.`attachments`
2017-03-27 11:15:24 INFO connection validated on 127.0.0.1:3306
2017-03-27 11:15:24 INFO User has ALL privileges
2017-03-27 11:15:24 INFO binary logs validated on 127.0.0.1:3306
2017-03-27 11:15:24 INFO Restarting replication on 127.0.0.1:3306 to make sure binlog settings apply to replication thread
2017-03-27 11:15:24 INFO Inspector initiated on PVGM50860507A:3306, version 10.2.4-MariaDB-log
2017-03-27 11:15:24 INFO Table found. Engine=InnoDB
2017-03-27 11:15:24 INFO Estimated number of rows via EXPLAIN: 17
2017-03-27 11:15:24 INFO Recursively searching for replication master
2017-03-27 11:15:24 INFO Master found to be PVGM50860507A:3306
2017-03-27 11:15:24 WARNING log_slave_updates not found on 127.0.0.1:3306, but executing directly on master, so I'm proceeeding
2017-03-27 11:15:24 INFO connection validated on 127.0.0.1:3306
2017/03/27 11:15:24 binlogsyncer.go:75: [info] create BinlogSyncer with config &{99999 mysql 127.0.0.1 3306 root   false false <nil>}
2017-03-27 11:15:24 INFO Connecting binlog streamer at mysql-bin.000002:18481
2017/03/27 11:15:24 binlogsyncer.go:241: [info] begin to sync binlog from position (mysql-bin.000002, 18481)
2017/03/27 11:15:24 binlogsyncer.go:134: [info] register slave for master server 127.0.0.1:3306
2017-03-27 11:15:24 INFO connection validated on 127.0.0.1:3306
2017-03-27 11:15:24 INFO connection validated on 127.0.0.1:3306
2017-03-27 11:15:24 INFO will use time_zone='SYSTEM' on applier
2017-03-27 11:15:24 INFO Examining table structure on applier
2017/03/27 11:15:24 binlogsyncer.go:568: [info] rotate to (mysql-bin.000002, 18481)
2017-03-27 11:15:24 INFO rotate to next log name: mysql-bin.000002
2017-03-27 11:15:25 INFO Applier initiated on PVGM50860507A:3306, version 10.2.4-MariaDB-log
2017-03-27 11:15:25 INFO Droppping table `excelsior_development`.`_attachments_gho`
2017-03-27 11:15:25 INFO Table dropped
2017-03-27 11:15:25 INFO Droppping table `excelsior_development`.`_attachments_ghc`
2017-03-27 11:15:25 INFO Table dropped
2017-03-27 11:15:25 INFO Creating changelog table `excelsior_development`.`_attachments_ghc`
2017-03-27 11:15:25 INFO Changelog table created
2017-03-27 11:15:25 INFO Creating ghost table `excelsior_development`.`_attachments_gho`
2017-03-27 11:15:25 INFO Ghost table created
2017-03-27 11:15:25 INFO Altering ghost table `excelsior_development`.`_attachments_gho`
2017-03-27 11:15:25 INFO Ghost table altered
2017-03-27 11:15:25 INFO Waiting for ghost table to be migrated. Current lag is 0s
2017-03-27 11:15:25 INFO Intercepted changelog state GhostTableMigrated
2017-03-27 11:15:25 INFO Handled changelog state GhostTableMigrated
2017-03-27 11:15:25 INFO Chosen shared unique key is PRIMARY
2017-03-27 11:15:25 INFO Shared columns are id,attachable_id,attachable_type,attachment,created_at,updated_at,creator_id,filesize,mime_type,md5_digest,type,origin,uuid,network_id
2017-03-27 11:15:25 INFO Listening on unix socket file: /tmp/gh-ost.excelsior_development.attachments.sock
2017-03-27 11:15:25 INFO Migration min values: [38]
2017-03-27 11:15:25 INFO Migration max values: [54]
2017-03-27 11:15:25 INFO Waiting for first throttle metrics to be collected
2017-03-27 11:15:25 INFO First throttle metrics collected
# Migrating `excelsior_development`.`attachments`; Ghost table is `excelsior_development`.`_attachments_gho`
# Migrating PVGM50860507A:3306; inspecting PVGM50860507A:3306; executing on PVGM50860507A
# Migration started at Mon Mar 27 11:15:24 +0800 2017
# chunk-size: 1000; max-lag-millis: 1500ms; max-load: ; critical-load: ; nice-ratio: 0.000000
# throttle-additional-flag-file: /tmp/gh-ost.throttle
# Serving on unix socket: /tmp/gh-ost.excelsior_development.attachments.sock
Copy: 0/17 0.0%; Applied: 0; Backlog: 0/100; Time: 0s(total), 0s(copy); streamer: mysql-bin.000002:22463; State: migrating; ETA: N/A
Copy: 0/17 0.0%; Applied: 0; Backlog: 0/100; Time: 1s(total), 1s(copy); streamer: mysql-bin.000002:28971; State: migrating; ETA: N/A
2017-03-27 11:15:26 INFO Row copy complete
Copy: 17/17 100.0%; Applied: 0; Backlog: 0/100; Time: 1s(total), 1s(copy); streamer: mysql-bin.000002:32512; State: migrating; ETA: due
2017-03-27 11:15:26 INFO Grabbing voluntary lock: gh-ost.32.lock
2017-03-27 11:15:26 INFO Setting LOCK timeout as 6 seconds
2017-03-27 11:15:26 INFO Looking for magic cut-over table
2017-03-27 11:15:26 INFO Creating magic cut-over table `excelsior_development`.`_attachments_del`
2017-03-27 11:15:26 INFO Magic cut-over table created
2017-03-27 11:15:26 INFO Locking `excelsior_development`.`attachments`, `excelsior_development`.`_attachments_del`
2017-03-27 11:15:26 INFO Tables locked
2017-03-27 11:15:26 INFO Session locking original & magic tables is 32
2017-03-27 11:15:26 INFO Writing changelog state: AllEventsUpToLockProcessed:1490584526400170764
2017-03-27 11:15:26 INFO Intercepted changelog state AllEventsUpToLockProcessed
2017-03-27 11:15:26 INFO Handled changelog state AllEventsUpToLockProcessed
2017-03-27 11:15:26 INFO Waiting for events up to lock
Copy: 17/17 100.0%; Applied: 0; Backlog: 1/100; Time: 2s(total), 1s(copy); streamer: mysql-bin.000002:40449; State: migrating; ETA: due
2017-03-27 11:15:27 INFO Waiting for events up to lock: got AllEventsUpToLockProcessed:1490584526400170764
2017-03-27 11:15:27 INFO Done waiting for events up to lock; duration=967.503825ms
# Migrating `excelsior_development`.`attachments`; Ghost table is `excelsior_development`.`_attachments_gho`
# Migrating PVGM50860507A:3306; inspecting PVGM50860507A:3306; executing on PVGM50860507A
# Migration started at Mon Mar 27 11:15:24 +0800 2017
# chunk-size: 1000; max-lag-millis: 1500ms; max-load: ; critical-load: ; nice-ratio: 0.000000
# throttle-additional-flag-file: /tmp/gh-ost.throttle
# Serving on unix socket: /tmp/gh-ost.excelsior_development.attachments.sock
Copy: 17/17 100.0%; Applied: 0; Backlog: 0/100; Time: 2s(total), 1s(copy); streamer: mysql-bin.000002:41207; State: migrating; ETA: due
2017-03-27 11:15:27 INFO Setting RENAME timeout as 3 seconds
2017-03-27 11:15:27 INFO Session renaming tables is 33
2017-03-27 11:15:27 INFO Issuing and expecting this to block: rename /* gh-ost */ table `excelsior_development`.`attachments` to `excelsior_development`.`_attachments_del`, `excelsior_development`.`_attachments_gho` to `excelsior_development`.`attachments`
2017-03-27 11:15:27 INFO Found atomic RENAME to be blocking, as expected. Double checking the lock is still in place (though I don't strictly have to)
2017-03-27 11:15:27 INFO Checking session lock: gh-ost.32.lock
2017-03-27 11:15:27 INFO Connection holding lock on original table still exists
2017-03-27 11:15:27 INFO Will now proceed to drop magic table and unlock tables
2017-03-27 11:15:27 INFO Dropping magic cut-over table
2017-03-27 11:15:27 INFO Releasing lock from `excelsior_development`.`attachments`, `excelsior_development`.`_attachments_del`
2017-03-27 11:15:27 INFO Tables unlocked
2017-03-27 11:15:27 INFO Tables renamed
2017-03-27 11:15:27 INFO Lock & rename duration: 1.027891832s. During this time, queries on `attachments` were blocked
2017-03-27 11:15:27 INFO Looking for magic cut-over table
2017/03/27 11:15:27 binlogsyncer.go:102: [info] syncer is closing...
2017/03/27 11:15:27 binlogstreamer.go:47: [error] close sync with err: sync is been closing...
2017/03/27 11:15:27 binlogsyncer.go:117: [info] syncer is closed
2017-03-27 11:15:27 INFO Closed streamer connection. err=<nil>
2017-03-27 11:15:27 INFO Droppping table `excelsior_development`.`_attachments_ghc`
2017-03-27 11:15:27 INFO Table dropped
2017-03-27 11:15:27 INFO Am not dropping old table because I want this operation to be as live as possible. If you insist I should do it, please add `--ok-to-drop-table` next time. But I prefer you do not. To drop the old table, issue:
2017-03-27 11:15:27 INFO -- drop table `excelsior_development`.`_attachments_del`
2017-03-27 11:15:27 INFO Done migrating `excelsior_development`.`attachments`
2017-03-27 11:15:27 INFO Removing socket file: /tmp/gh-ost.excelsior_development.attachments.sock
# Done
```

DDL of the table:
```
CREATE TABLE `attachments` (
  `id` int(11) NOT NULL,
  `attachable_id` int(11) DEFAULT NULL,
  `attachable_type` varchar(191) DEFAULT NULL,
  `attachment` varchar(191) DEFAULT NULL,
  `created_at` datetime NOT NULL,
  `creator_id` int(11) DEFAULT NULL,
  `filesize` int(11) DEFAULT 0,
  `md5_digest` varchar(191) DEFAULT NULL,
  `mime_type` varchar(191) DEFAULT NULL,
  `network_id` int(11) DEFAULT NULL,
  `origin` int(11) NOT NULL DEFAULT 0,
  `type` varchar(30) DEFAULT NULL,
  `updated_at` datetime NOT NULL,
  `uuid` varchar(191) DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `idx_attachable_type_attachable_id` (`attachable_type`,`attachable_id`) USING BTREE,
  KEY `index_attachments_on_creator_id` (`creator_id`) USING BTREE
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

**Comment by shlomi-noach** [comment]
duplicate of https://github.com/github/gh-ost/issues/383

TL;DR: there is no problem with the migration; the data continues to be safe and `gh-ost` is actually unaware of the error at this time. An upgrade of a more recent `go-mysql` caused `gh-ost` to attempt closing the replication/streamer connection twice, hence the error.

I understand and agree that this message appearing causes concerns and will work to resolve this.

**Comment by shlomi-noach** [comment]
Addressed by https://github.com/github/gh-ost/pull/389

**Comment by ChinaVogel** [comment]
@shlomi-noach thanks for the fast response && fix!

So I can move on with more tests :)

**Comment by shlomi-noach** [comment]
fixed by https://github.com/github/gh-ost/pull/389 (no release yet)

**Comment by shravanSD** [comment]
Hello,

Are you sure you specified the "--execute" ?

I have had the issue when I did not specify --execute.
